net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,186 @@
+	ac := Endpoint{"http://login.tailscale.com/generate_204", http.StatusNoContent, "", true, Tailscale}
+	endpoints = append(endpoints, cs, ac)
+
+	// Lastly, to be safe, let's also include some well-known captive portal detection URLs that are not under the
+	// tailscale.com umbrella. These are less likely to be blocked on public networks since blocking them
+	// would break captive portal detection for many devices.
+	switch goos {
+	case "windows":
+		endpoints = append(endpoints, Endpoint{"http://www.msftconnecttest.com/connecttest.txt", http.StatusOK, "Microsoft Connect Test", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.msftncsi.com/ncsi.txt", http.StatusOK, "Microsoft NCSI", false, Platform})
+	case "darwin", "ios":
+		endpoints = append(endpoints, Endpoint{"http://captive.apple.com/hotspot-detect.html", http.StatusOK, "Success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.thinkdifferent.us/", http.StatusOK, "Success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://www.airport.us/", http.StatusOK, "Success", false, Platform})
+	case "android":
+		endpoints = append(endpoints, Endpoint{"http://connectivitycheck.android.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://connectivitycheck.gstatic.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://play.googleapis.com/generate_204", http.StatusNoContent, "", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://clients3.google.com/generate_204", http.StatusNoContent, "", false, Platform})
+	default:
+		endpoints = append(endpoints, Endpoint{"http://detectportal.firefox.com/success.txt", http.StatusOK, "success", false, Platform})
+		endpoints = append(endpoints, Endpoint{"http://network-test.debian.org/nm", http.StatusOK, "NetworkManager is online", false, Platform})
+	}
+
+	// Sort the endpoints by provider so that we can prioritize DERP nodes in the preferred region, followed by
+	// Tailscale endpoints, followed by platform-specific endpoints, then DERP nodes in any other region.
+	slices.SortFunc(endpoints, func(x, y Endpoint) int {
+		return cmp.Compare(x.Provider, y.Provider)
+	})
+
+	return endpoints
+}
+
+// responseLooksLikeCaptive checks if the given HTTP response matches the expected response for the Endpoint.
+func (e Endpoint) responseLooksLikeCaptive(r *http.Response, logf logger.Logf) bool {
+	defer r.Body.Close()
+
+	// Check the status code first.
+	if r.StatusCode != e.StatusCode {
+		logf("unexpected status code in captive portal response: want=%d, got=%d", e.StatusCode, r.StatusCode)
+		return true
+	}
+
+	// If we don't have an expected content string, we don't need to check the response body.
+	if e.ExpectedContent == "" {
+		logf("captive portal check response: no need to check response body for %q", e.URL)
+		return false
+	}
```

Comment:
I think this means that we don't ever check the Tailscale challenge response, right? Since those endpoints have no expected content?

Reply:
Oh damn, nice catch!

Reply:
Fixed